### PR TITLE
Implement Layer Normalization

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

**Comment by ShikharJ** [comment]
@rcurtin @zoq Can I have a review of this as well?

File: src/mlpack/methods/ann/layer/layer_norm_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void LayerNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  const arma::mat inputMean = input.each_col() - mean;
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // Step 1: dl / dxhat *
```
[the remark this answers is not in the document]
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: Yeah, I kind of had that doubt as well. I'll get back to you regarding this.

**Comment by zoq** [comment]
Let me know if I should fix the merge issue.

File: src/mlpack/methods/ann/layer/layer_norm_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: @rcurtin What is your take on this? `Layer Normalization` follows directly from `Batch Normalization` (with `Deterministic` set to `false` at all times), so shouldn't the procedure for calculating `Gradients` also be the same?

[the remark this answers is not in the document]
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: @zoq I couldn't find any definitive resource on this. The output matches the expected one though.

**Comment by ShikharJ** [comment]
@zoq Adding the test for complex network would be redundant,  as we'll be directly using this for the `WGAN` tests. Maybe we can add a test there, since `Layer Normalization` isn't that frequently used for CNNs?

**Comment by zoq** [comment]
Good point, in this case, let us write another test which incorporates the layer in a more complex network later.

File: src/mlpack/methods/ann/layer/layer_norm_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zoq) [reply src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: They are the same, just wanted to avoid the confusion that step #number doesn't match with the number in the paper, perhaps removing the number from the step is all we need.

[the remark this answers is not in the document]
Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: Hmm, updated it. Now it's ready for merging.

**Comment by ShikharJ** [comment]
@zoq Yes, this is ready to be merged.

File: src/mlpack/methods/ann/layer/layer_norm.hpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Declaration of the Layer Normalization class. The layer transforms
+ * the input data into zero mean and unit variance and then scales and shifts
+ * the data by parameters, gamma and beta respectively over a single training
+ * data. These parameters are learnt by the network.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/layer_norm.hpp]: If you like, it might be useful to add a comment here differentiating 'layer normalization' from 'batch normalization'---I imagine many users will be familiar with the terms, but not have any insight on why to use which one.  So if you can think of something to add here, it could be useful, otherwise I think it is not a huge deal. :)

**Comment by zoq** [comment]
Thanks, really productive week 👍
